### PR TITLE
feature: Add Permit to BridgeToken

### LINKED ISSUE #480: feature: add permit to the standard bridge token
use uniswap-style permit function for eip712 allowances

### PR DESCRIPTION
**Pull request by prestwich**
closes #480 

drive-by changes:
- running prettier in optics-deploy
- changing some for-let loops into maps

### CODE REVIEW

**Comment by prestwich** [comment]
This is implemented and tested. Ready for review

File: solidity/optics-xapps/contracts/bridge/BridgeToken.sol [hunk 1]
```diff
@@ -91,4 +98,55 @@ contract BridgeToken is IBridgeToken, Ownable, ERC20 {
...
+        bytes32 hashStruct = keccak256(
+            abi.encode(_PERMIT_TYPEHASH, owner, spender, value, nonce, deadline)
+        );
+
+        bytes32 digest = keccak256(
+            abi.encodePacked(uint16(0x1901), domainSeparator(), hashStruct)
+        );
+
+        address signer = ecrecover(digest, v, r, s);
+        require(signer == owner, "ERC20Permit: invalid signature");
```
Comment (asaj) [inline solidity/optics-xapps/contracts/bridge/BridgeToken.sol:147]: Check signer != 0?

**Review by asaj — APPROVED** [review]
LGTM

File: solidity/optics-xapps/contracts/bridge/BridgeToken.sol (hunk 1 above)
Comment (prestwich) [reply solidity/optics-xapps/contracts/bridge/BridgeToken.sol:147]: good call

Comment (prestwich) [reply solidity/optics-xapps/contracts/bridge/BridgeToken.sol:148]: oh to follow up on this, it isn't _technically_ required, as that is checked in the `_approve` function. however, I prefer to have `permit` explicitly check all of its invariants without relying on implementation details of other functions

**Comment by prestwich** [comment]
@anna-carroll this is ready for review/merge

File: solidity/optics-xapps/contracts/bridge/BridgeToken.sol [hunk 2]
```diff
@@ -91,4 +98,56 @@ contract BridgeToken is IBridgeToken, Ownable, ERC20 {
...
+        require(owner != address(0), "ERC20Permit: owner zero address");
+
+        uint256 nonce = nonces[owner];
+
+        bytes32 hashStruct = keccak256(
+            abi.encode(_PERMIT_TYPEHASH, owner, spender, value, nonce, deadline)
+        );
+
+        bytes32 digest = keccak256(
+            abi.encodePacked(uint16(0x1901), domainSeparator(), hashStruct)
```
Comment (anna-carroll) [inline solidity/optics-xapps/contracts/bridge/BridgeToken.sol:144]: `0x19` is the [EIP 712](https://eips.ethereum.org/EIPS/eip-712) prefix & `0x01` is [EIP 191](https://eips.ethereum.org/EIPS/eip-191) version byte?

File: solidity/optics-xapps/contracts/bridge/BridgeToken.sol [hunk 3]
```diff
@@ -91,4 +98,56 @@ contract BridgeToken is IBridgeToken, Ownable, ERC20 {
...
+            chainId := chainid()
+        }
+        return
+            keccak256(
+                abi.encode(
+                    keccak256(
+                        "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
+                    ),
+                    keccak256(bytes(token.name)),
+                    keccak256(bytes("1")),
```
Comment (anna-carroll) [inline solidity/optics-xapps/contracts/bridge/BridgeToken.sol:118]: should we make this a 
```
string public immutable _EIP712_STRUCTURED_DATA_VERSION = "1";
```
so it's less of a magic number for the un-initiated to read?

**Review by anna-carroll — COMMENTED** [review]
made a couple changes at https://github.com/celo-org/optics-monorepo/pull/510/commits/2066e405cdbcac0984770e288a1631fa7cdaae27
- re-arrange functions, rename memory variables acc. to style guide
- document the permit function
- remove whitespace from Solidity functions
- make magic numbers calculated at runtime into immutable variables

**Review by anna-carroll — DISMISSED** [review]
LGTM - nice on writing tests 💯 

@prestwich can you make sure that the naming for the immutable variables that I introduced at https://github.com/celo-org/optics-monorepo/pull/510/commits/2066e405cdbcac0984770e288a1631fa7cdaae27 is accurate for what these variables are doing?

**Review by anna-carroll — APPROVED** [review]
Rust lint is failing, but going to approve and force-merge anyway cc @emberian

**Comment by emberian** [comment]
oh no! taking a look

**Pull request merged**